File: Snippets/Union-Find/union_find.py
def find(x):
    while x != parents[x]:
        parents[x] = parents[parents[x]]
        x = parents[x]
    return x


def union(x, y):
    x = find(x)
    y = find(y)
    if x != y:
        parents[y] = x
        return True
    else:
        return False


N = 1000
parents = [i for i in range(N+1)]

File: Snippets/Union-Find/UnionFind.kt
import java.io.*

var parent = IntArray(0)
var values = IntArray(0)
fun find(x: Int): Int {
    var i = x
    while (i != parent[i]) {
        parent[i] = parent[parent[i]]
        i = parent[i]
    }
    return i;
}

fun union(x: Int, y: Int): Boolean {
    val i = find(x)
    val j = find(y)
    if (i == j) return false
    if (values[i] < values[j]) {
        parent[i] = j
        values[j] += values[i]
    } else {
        parent[j] = i
        values[i] += values[j]
    }
    return true
}

var br = BufferedReader(InputStreamReader(System.`in`))
fun main() {

}